clab/config/template.go

```diff
@@ -0,0 +1,203 @@
+package config
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/srl-labs/containerlab/types"
+)
+
+type ConfigSnippet struct {
+	TargetNode *types.Node
+	// the Rendered template
+	Data []byte
+	// some info for tracing/debugging
+	templateName, source string
+	// All the variables used to render the template
+	vars *map[string]string
+}
+
+// internal template cache
+var templates map[string]*template.Template
+
+func LoadTemplate(kind, templatePath string) error {
+	if templates == nil {
+		templates = make(map[string]*template.Template)
+	}
+	if _, ok := templates[kind]; ok {
+		return nil
+	}
+
+	tp := filepath.Join(templatePath, kind, "*.tmpl")
```

Comment:
I think we shouldn't enforce dir structure by asking `kind` to be a subdirectory
IMHO it is better to just use `templatePath` and `*.tmpl`

Reply:
If the kind is not included you cannot build an interoperating lab. In the example I build connectivity & IGP between SR-OS and SRL.

Not having the kind in the path or filename means you'll need it in the template itself, pass a kind variable, but the template will become overly complex...

Reply:
When I started to think about it more, that is the flow I got pictured in my head:

1. lets imagine the goal is the same -  to configure a basic ip connectivity between `srl1` and `sr1` nodes
2. In my view, the natural way of handling a task like that is to create a template (or a set of templates) and let the user to decide if the template(s) should be per-node, per-kind or global defaults.
3. by building on top of this idea, I thought that is how I would have done it:
    a user creates a template for srl named `~/srl.tmpl` and for sr `~/sr.tmpl`. At this moment the templates files belong to the same directory. 
2. then in the clab file the user can specify the paths to the templates which are used for a particular node, smth like this:
    ```yaml
    nodes:
      srl1:
        config:
          templates:
            - ~/srl.tmpl
    ```
     this definition will tell clab to use `~/srl.tmpl` for this node (and `sr.tmpl` for sros accordingly)
3. Note how here I pass the paths to the templates that should be used for this particular node. I could also create a `config` section on the kind level with the same path and that would have the effect of using the provided template for all SRL kinds.
3. This approach assumes that `--template-path` [flag](https://github.com/srl-labs/containerlab/pull/383/files/2a8df0115d1e284ea4889beec5f4f32fe9e8d8b3) is not really needed, as the paths to the templates are passed in the clab file
4. We can also add glob support in the provided paths, i.e. 
    ```yaml
    nodes:
      srl1:
        config:
          templates:
            - ~/srl*.tmpl
    ```
    and that will parse all the templates prefixed with `srl` string.


### Why?
Now to the why I think this is a more powerful approach.

1. we are not limiting users or dictate the rules as to how the structure of the template files should look like. They can keep the files anywhere they like and have the flexibility on selecting which templates will be in use
2. the clab file stores the paths to the templates and that makes configuration even easier. You don't need to have provide the template path in the cli, as this information is stored in the clab file itself and it is clear which template will be used for which nodes
3. by scoping the templates selection all the way to a per-node basis we have the best flexibility in templates application. Some nodes (even of the same kind) can use different templates if a user wants that

Reply:
But templates is not a fire-and-forget thing. It is incremental. Incremental enables us (slow) humans to test templates, capture output, etc. Templates in my vision was also more than config, it's a whole way of automating workflow across an entire lab (and this go implementation is lightning fast, almost like typing CLI!)

As an example I ran this lab last week:
```
# 1. Clear all config
# 2. Apply base - router/port/ISIS
# 3. apply seamless for ABRs, export system address etc
# 4.apply VPRN and set VPRN loopback address
clab_ config -t rrfwd2as.clab.yml -p $TEMPLATES1 -l delete,base,seamless,vprn

# Fix the ASBR
clab_ config -t rrfwd2as.clab.yml -p . -l rrfwd2as

# Show route table
clab_ config -t rrfwd2as.clab.yml -p $TEMPLATES1 -l show-route-table

# Hunt around
clab_ config -t rrfwd2as.clab.yml -p . -l show 

clab_ config -t rrfwd2as.clab.yml -p . -l set-rr-fwd

clab_ config -t rrfwd2as.clab.yml -p . -l show
```

As an example `delete` is a nice utility to make sure you captured all the changes in the layers of templates. Very powerful to restart this iterative process while building and testing my configs: `clab_ config -t rrfwd2as.clab.yml -p $TEMPLATES1 -l delete`

The `vr-sros/delete-node.tmpl` template contains:
```
delete /configure router
delete /configure policy-options
delete /configure service
```

I'll mail you the result as powerpoint for reference

Reply:
fair point, it makes sense.
what then about this functionality -> currently you assume that the templates are aggregated after kinds. So that template files are kept under the `kind/template.tmpl` path.

I think in the future what would make sense is to use labels to tell the templating feature which base paths to use when configuring the nodes. Consider an example where you have a leaf-spine pop and you want to create templates for leafs and spines. Both of these network functions are of SRL kind, and that will make it impossible to create different templates for those nodes.
If, on the other hand, we say that templating engine can look for a base path that equals to some label, then it will make possible to flexibly create a folder structure as user sees fit.

In my example above, the file structure can be as follows:

```
.
├── config
│   ├── leaf
│   │   └── base.tmpl
│   └── spine
│       └── base.tmpl
└── mylab.clab.yml

3 directories, 3 files
```

and my clab file will have a label for each node
```yaml
leaf:
  kind: srl
  labels:
    role: leaf
spine:
  kind: srl
  labels:
    role: spine
```

and then `config` command will have a `--label role` flag to tell templating feature where to find the template files

wdyt?

Reply:
you can easily do that today:

```
{{ if eq .role "leaf" }}
do leafy stuff
{{ end }}

{{ if eq .role "spine" }}
do spiny stuff
{{ end }}
```

What you are suggesting here and in the next comment is a contradiction ;-)

- here: Add another role based file resolution indirection 
- next: Remove a file resolution based indirection in favour of a var (we'll end up with `links := []map`)

btw, I like your suggestion below, since I end up doing many more empty -link templates than not.

Reply:
If you really want the role base, I would suggest `/config/<kind>/base-<role>.tmpl`, but would prefer not to do that initially.

So this is a generalization of today's -node & -link (which I am going to remove now...)

Reply:
After some more thought, we could combine these ideas:

<pre>/templatePath/<b>{name}</b>-<b>{role or kind}</b>.tmpl</pre>

So this gives you:
(1) a flat template folder
(2) keeps the *kind* based config as it is today
(3) allow you to overwrite the default *kind* by specifying a *role*

The shows & debug is quite explicit in what it uses, so should be easy to follow.

The only trick now is if you have different devices AND want to use roles, you need to add the kind/some id. So a SROS `role: leaf` and a SRL `role: srleaf`, but that seem pretty acceptable.

So you can also see this as the default role being the kind.

Reply:
> /templatePath/{name}-{role or kind}.tmpl

I like that. Maybe let's take this as an agreed file-search logic and let it out in the wild to see how this will be perceived and what are the limitations of it